tools/Google.Cloud.Tools.TrimCoverageReport/Program.cs

```diff
@@ -0,0 +1,124 @@
+
+            var fileElements = document.Root.Element("FileIndices").Elements("File");
+            var generatedFiles = fileElements.Where(file => IsGeneratedFile(file.Attribute("Name").Value)).ToList();
+            var generatedIndices = new HashSet<int>(generatedFiles.Select(file => (int) file.Attribute("Index")));
+            generatedFiles.Remove();
+
+            document.Descendants("Statement")
+                .Where(st => generatedIndices.Contains((int)st.Attribute("FileIndex")))
+                .Remove();
+
+            RecomputeAndSimplify(document, "Method", "Statement", GetStatementCoverage);
+            RecomputeAndSimplify(document, "Type", "Method", GetContainerCoverage);
+            RecomputeAndSimplify(document, "Namespace", "Type", GetContainerCoverage);
+            RecomputeAndSimplify(document, "Assembly", "Namespace", GetContainerCoverage);
+        }
+
+        private static void RecomputeAndSimplify(XDocument document, string parentElement, string descendantElement, Func<XElement, (int statements, int covered)> coverageSelector)
+        {
+            List<XElement> elementsToRemove = new List<XElement>();
+            foreach (var parent in document.Descendants(parentElement))
+            {
+                var coverage = parent
+                    .Descendants(descendantElement)
+                    .Select(coverageSelector)
+                    .Aggregate((statements: 0, covered: 0), (current, next) => ((current.statements + next.statements), (current.covered + next.covered)));
+                if (coverage.statements == 0)
+                {
+                    elementsToRemove.Add(parent);
+                }
+                else
+                {
+                    parent.SetAttributeValue(TotalStatements, coverage.statements);
+                    parent.SetAttributeValue(CoveredStatements, coverage.covered);
+                    parent.SetAttributeValue(CoveragePercent, (coverage.covered * 100) / coverage.statements);
+                }
+            }
+
+            // Remove all the elements with 0 statements.
+            elementsToRemove.Remove();
+        }
+
+        private static (int statements, int covered) GetStatementCoverage(XElement statement) =>
+            (1, (bool)statement.Attribute("Covered") ? 1 : 0);
+
+        private static (int statements, int covered) GetContainerCoverage(XElement container) =>
+            ((int)container.Attribute(TotalStatements), (int)container.Attribute(CoveredStatements));
+
+        /// <summary>
+        /// Detects whether a file is generated. All three code generators we use (protoc, grpc,
+        /// toolkit) include a comment starting with "// Generated" early in the file.
+        /// </summary>
+        /// <param name="path"></param>
+        /// <returns></returns>
+        static bool IsGeneratedFile(string path) =>
+            File.ReadLines(path).Take(20).Any(line => line.StartsWith("// Generated"));
```

Comment:
I think it'll be fine - it's true for all the generators we have now, and we'll soon spot if we introduce a new generator, I think.